miracle/core/extractors.py

```diff
@@ -0,0 +1,148 @@
+import os
+import pyunpack as p
+import os.path as path
+import shutil as s
+import collections as c
+
+from .metadata import load_metadata, AnalysisMetadata
+from ..settings import MIRACLE_ANALYSIS_DIRECTORY
+
+
+class MetaDataImportError(Exception): pass
+
+
+def fileext_gen(file_names):
+    for file_name in file_names:
+        yield os.path.splitext(file_name)
+
+class SHPFileGlob(object):
+    """
+    Shapefiles should be thought of as one unit
+
+    e.g. A shapefile is composed of at least 4 files:
+     a .shp, a .dbf, a .prj and a .shx
+    """
+    REQUIRED_EXTENSIONS = {".shp", ".dbf", ".prj", ".shx"}
+
+    @staticmethod
+    def from_files(file_names, log):
+
+        shp_names = []
+        for name, ext in fileext_gen(file_names):
+            if ext == ".shp":
+                shp_names.append(name)
+
+        shp_file_groups = c.defaultdict(set)
+        for name, ext in fileext_gen(file_names):
+            if name in shp_names:
+                shp_file_groups[name].add(ext)
+
+        names = []
+        for shp_name in shp_names:
+            if SHPFileGlob.is_valid(shp_file_groups[shp_name]):
+                names.append(shp_name + ".shp")
+            else:
+                log.append("Shapefile is missing extensions {0} (only has {1})".
+                           format(shp_name, SHPFileGlob.REQUIRED_EXTENSIONS.difference(shp_file_groups[shp_name])))
+
+            for ext in shp_file_groups[shp_name]:
+                file_names.remove(shp_name + ext)
+
+        return names
+
+    @staticmethod
+    def is_valid(extensions):
+        return extensions.issuperset(SHPFileGlob.REQUIRED_EXTENSIONS)
+
+
+class OtherFileGlob(object):
+    @staticmethod
+    def from_files(file_names, log):
+        return file_names
+
+
+class FileGlobber(object):
+    """
+    Groups files that belong together and returns an iterator
+    """
+    @staticmethod
+    def extract_metadata(analysis_path):
+        walker = os.walk(analysis_path)
+
+        metadata = []
+        log = []
+        for root, dirs, files in walker:
+            files = [f for f in files if not f[0] == '.']
+            dirs[:] = [d for d in dirs if not d[0] == '.']
+            shp_files = SHPFileGlob.from_files(files, log)
+            other_files = OtherFileGlob.from_files(files, log)
+
+            all_files = other_files + shp_files
+            for file in all_files:
+                fullpath = path.join(root, file)
+                metadata.append(load_metadata(fullpath))
+
+        return AnalysisMetadata(metadata, log)
+
+class Extractor(object):
```

Comment:
As the entry point into metadata extraction, clear docs would be nice on expected usage and API (e.g., do we create one Extractor per dataset upload or should we always use the static `from_archive` factory method to construct Extractors?).